Rp2040

ecarjat opened the pull request:
This PR enables the code ModbusServerRTU to run on a RP2040 using the arduino framework and the earlephilhower core.
I have only tested the RTU server part. The client RTU replicates the changes done in server and should work as is but it has not been tested.

Miq1 commented:
_Nice!_
I have no RP2040 yet to test it, so it will have to wait for that.
I have a few objections, though:
- the addition of ``__linux__`` to the ``#ifdef`` checks in the async TCP code is not sensible in my opinion, as there will be no such support without completely rewriting the code.
- your code formatting preferences are different to ours, so all the replaced braces are irritating at least. I would ask you to maintain our formatting habits instead.

ecarjat commented:
Thanks for the quick reply.

- On code formatting, I did change it so that the braces match your formatting habits. This is what commit `[2824eff]` intends to do. My current diff does not show a difference. Could you tell me in which file you see a difference ?
- On `__linux__` I've added all the platforms previously listed in the `options.h` file without checking if this would have worked on the platform. Happy to change it to `defined(ESP32) || defined(ESP8266)` if that that's the only platforms it works for. 
- On `#if HAS_FREERTOS`, on second thought I believe this should be changed to `#if ESP8266 || ESP32` as the code in this function is specific to these platforms.

Miq1 commented:
My bad, sorry! I had a look only at the first commit.

Regarding those RTOS task calls, that could be valid for the RP2040 as well - IIRC there are two cores, too? On the ESP32 running the worker task on a dedicated core helps to maintain the bus speed at higher baud rates, the ESP32 can go up to 5Mbaud with it. The RTU server or client do not run on the ESP8266 for exactly the reason of having only one core.

ecarjat commented:
The RP2040 has two cores indeed, however the function used `xTaskCreatePinnedToCore` is proprietary to Expressif. According to [this](https://forums.freertos.org/t/why-was-xtaskcreatepinnedtocore-removed-from-the-latest-release/15766), the "proper" RTOS way to do it is to set affinity. To mimic `xTaskCreatePinnedToCore` functionality I had to split it into two calls.

Miq1 commented:
We are getting there. 😀 
To swap out the Espressif task calls you are using the ARDUINO_ARCH... define. I would think the HAS_FREERTOS2040 would be more consistent instead. At one spot of these there is a isolated ``#elsif`` - shouldn't that be an ``#else``?
Finally the addition of ``__linux__`` still is in the async TCP files...

I will order a RP2040 for me to test. I would like it better to have the TCP parts also working on that MCU. Did you try those as well? I suppose your definition in ``options.h`` is protecting you from all the ``HAS_FREERTOS`` encapsulated code.

Miq1 commented:
Excellent! Thank you so much. I expect to get the RP2040 on Saturday, so I may approve the PR beginning of next week.

Miq1 commented:
What are your recent changes about to achieve? Just renaming the config variable? If so, why is there no matching commit for ``options.h``?

ecarjat commented:
Just revisiting this project and cleaning up code on my end. Did not think about the side effect on this PR. I can remove it for the time being if you prefer.

Miq1 commented on src/options.h:
I am sceptical about the ``do ... while(0);`` construct. What is it good for?

ecarjat replied:
The do { ... } while (0) wrapper is there to make EMODBUS_SET_TASK_AFFINITY behave as a single statement macro, so it is safe in all call sites (especially if/else without braces) and does not leak scope. That said, in our current usage it is not strictly required, and I can simplify it if you prefer readability over the defensive macro pattern. I’m preparing a cleanup pass on this PR anyway, so I can either keep it with a short comment or switch to a simpler form.

Miq1 commented on src/RTUutils.h:
Setting the buffer size may not be relevant at all for the RP2040. It was invented on the ESP32 to get around a shortcoming in the UART handling of the arduinoesp32 core.

Commit:
rp2040: clean port for RTU support and FreeRTOS compatibility

- squash rp2040 branch history onto upstream/master

- include RP2040 FreeRTOS guards and RTU task affinity compatibility

- keep RP2040 RTU server example and library metadata updates

Miq1 commented:
@ecarjat Thank you for your contribution.
